### ISSUE TITLE
Widgets could form a DAG instead of a tree, and the "Widget tree" is actually the Element tree

### ISSUE THREAD

**Issue by liyuqian** [flutter/flutter#58661]
The same Widget instance can be added multiple times (see https://api.flutter.dev/flutter/widgets/Widget-class.html) so a Flutter app's widget hierarchy is actually a directed acyclic graph (DAG) instead of a tree. That is: the same Widget node could have multiple incoming edges as well as outgoing edges where an edge is a parent-child relationship. (The acyclic part comes from the fact that a widget is final and immutable so there's a topological order defined by the order of widget constructions, and therefore edges can't form a loop.)

However, it seems that we extensively use the term "Widget tree" in our docs, articles, slides, and tool descriptions. See https://flutter.dev/docs/testing/code-debugging#widget-tree for an example.

After checking the source code, I find that "Widget tree dump" is actually calling [debugDumpApp][] which dumps the Element tree. (The Elements do form a tree as the same Widget at different locations will instantiate different Elements. See https://api.flutter.dev/flutter/widgets/Element-class.html.)

This also happens in `flutter run` tool where `?` prints out `w Dump widget hierarchy to the console`. The `w` actually calls [debugDumpApp][] to print out the element tree instead of the widget DAG.

I wonder if we should clarify this so people get less confused, and are aware that they can reuse the same Widget instance multiple times in the hierarchy (which might be good for performance). Or is the Element so close to the Widget that it's Ok to just use the Widget tree to reference the Element tree?

[debugDumpApp]: https://github.com/flutter/flutter/blob/28138ec5e3271ae35295a46bf2eeaf554eeb2535/packages/flutter/lib/src/widgets/binding.dart#L977

**Comment by liyuqian** [comment]
CC @Hixie @sfshaza2 @filiph, @jonahwilliams, and @zanderso for thoughts on API docs, flutter.dev articles, and Flutter tools help info.

**Comment by Hixie** [comment]
Yeah, the "Widget tree" is really the "element tree". We feel comfortable calling it the widget tree because the widgets are how the elements are configured. It's a tree where each node is an element whose meaning is given by a widget.

"Widget tree" just rolls off the tongue more easily. :-)

**Comment by liyuqian** [comment]
Closing the issue as using the widget tree to reference the element tree is fine.
